### PR TITLE
[FLINK-4113] [runtime] Always copy first value in ChainedAllReduceDriver

### PR DESCRIPTION
**Pull request by greghogan**

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/chaining/ChainedAllReduceDriver.java [hunk 1]
```diff
@@ -89,7 +89,7 @@ public void collect(IT record) {
 		numRecordsIn.inc();
 		try {
 			if (base == null) {
-				base = objectReuseEnabled ? record : serializer.copy(record);
+				base = serializer.copy(record);
 			} else {
 				base = objectReuseEnabled ? reducer.reduce(base, record) : serializer.copy(reducer.reduce(base, record));
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/chaining/ChainedAllReduceDriver.java]: What if we return the `record` as the result of `reducer.reduce` (with `objectReuseEnabled == true`)? That would also be problematic, right?

**Comment by tillrohrmann** [comment]
Good fix @greghogan. We should definitely include it in the upcoming release. Would it be possible to write a test to guard the fix from future changes?

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/chaining/ChainedAllReduceDriver.java (hunk 1 above)
Comment (greghogan) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/chaining/ChainedAllReduceDriver.java]: We fixed this in FLINK-3340 for non-chained reduce drivers (where the driver chooses the object to deserialize into) but for chained drivers we cannot prevent one UDF from overwriting an object from a previous UDF. If you look in {{OverwriteObjects.java}} you will see {{testReduce}} fail.

**Comment by greghogan** [comment]
@tillrohrmann just added a test for `ChainedAllReduceDriver`. I didn't see a way to build a lightweight test for chained drivers as we do with the unchained drivers.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/chaining/ChainedAllReduceDriver.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/chaining/ChainedAllReduceDriver.java]: Alright. I guess users then have to live with that if they enable object reuse.

**Comment by tillrohrmann** [comment]
@greghogan the tests looks good. Thanks for your work :-) The failing test cases are unrelated. Will merge this PR.

**Pull request closed without merging**